### DISCUSSION TITLE
Question about optimisation of interpolated potentials

### DISCUSSION THREAD

**Discussion by smandhai** [jobovy/galpy#557] [Q&A]
Hi again! 

I'm having an issue of sorts where the orbits I run using an interpolated potential are running extremely slow (at least for my purposes). 10 orbits within the following potential for example take around 5 minutes. I was wondering if there are any tips or optimisations that I can make/if I'm doing this correctly? I'm currently using a leapfrog integrator to process the orbits. Further context below. Currently using galpy==1.8.0 (1.8.1 has an issue with loading in the unpickled potentials, but this seemingly goes away with 1.8.2 - which I can't install on my current machine as my local pip installation can't find it. This is probably a local issue with my laptop as the issue goes away when I use a different machine.)

`spi= pot.InterpSnapshotRZPotential(data_cut.s,rgrid=(np.log10(1e-4),np.log10(1e3),101),logR=True,zgrid=(-60.,60.,101),interpepifreq=True,
	interpverticalfreq=True,interpPot=True,zsym=False,enable_c=True,numcores=4)`
	
Is the convention with the log grid correct here? I suspect as custom interpolated potentials are generally more detailed than fully analytical potentials, they require longer times to integrate orbits over - is this the case?
	
#**Context:**

I am aiming to distil a galaxy simulation (in this case AURIGA - built on GADGET) into a workable potential to calculate orbits via galpy. I have loaded in the galaxy box into **pynbody**, pre-processed it (cutting around the region of interest), and used galpy to produce a potential for the stellar disc using the snippet above (where data_cut.s is the trimmed stellar component of the trimmed data set). This works fine and I've pickled it using the method from the discussion in https://github.com/jobovy/galpy/discussions/507. I have since unpickled it and loaded it into a separate script for the orbits. The orbits themselves aren't anything fancy, as I just wanted to get the basic foundation functional. I compared the included MWPotential2014 against "spi" potential produced above. The former runs in <1s, whilst the latter can take up to 20-30s per orbit - with the same input parameters. Producing the rotation curve alone can also take some time.

For the DM and Gas components of the galaxy, I originally used a SCF profile for their potentials. However, when compared to the rotation curves produced directly in **pynbody**, they were off - which I guess was expected as the SCF reflected a Hernquist potential whilst the galaxy's DM for example was more akin to a NFW profile. I have since switched these two similar interpolations as the stellar case above. Stacking the potentials doesn't seem to change the speed of the orbits. The rotation curves are more akin to what pynbody reports too.

For reference: Each component has few x 10-100 x 10^6 particles that are being passed through the interpolation script.

**Comment by jobovy** [comment]
Hi!

I think I'll need some more info to be able to help. 

* How exactly are you integrating the orbits (i.e., what `method=` are you using? If you are using `method='leapfrog'` then that will be very slow, even `leapfrog_c` might be very slow; I don't think there is any reason not to use one of the fast and accurate integrators like `dop853_c`)
* Could you try setting up an interpRZPotential for `MWPotentia2014` using the same grids and then doing the orbit integration with that? Once the potential is interpolated, they should take equally long no matter what the original potential was, so that might help me to locally check the issue.
* I don't think this matters here, but if you're just doing orbit integration, you don't need to set `interpepifreq=True, interpverticalfreq=True,interpPot=True`.
* Weird about the pickling not working in 1.8.1, but if it works again in 1.8.2 then I guess there is nothing to fix! Maybe the reason you can't upgrade is that your Python version is < 3.8? galpy 1.8.2 requires Python >= 3.8.


could you try just setting up an interpRZPotential with `MWPotential2014`

**Reply by smandhai** [reply]
Apologies for the delay! Thank you so much for the prompt response on your end, Jo! 

I'll respond to each point separately:

1. For the integration of the orbits, I cycled through the different methods and found leapfrog to be the fastest (at least for the debugging stage of things). The inputs are a time array and the interpolated potentials in a list (e.g. [stellar_potential,gas_potential, dm_potential]), the Orbit instance consists of an array of 10 vxvvs containing various different parameters. Changing from leapfrog to dop853 seemingly reduced the integration speed. Although, this isn't a certain statement as my python installation on my local machine (where I tested this) is acting up on the whole - at the time of writing this, I'm upgrading my installation. Will edit this with the outcome once that's done and I have retested. 

2. I tried this with the same binaries and grid, the integration was virtually instantaneous!

3. I mainly needed these to do some cross-checks between the interpolation that pynbody reports from the raw simulation and what I was getting out via GalPy. I did try to turn these off but couldn't seemingly calculate the rotation curve without them. On this note, are there alternative methods to calculate galaxy densities using the interpolated snapshots? 

4. I didn't know about the Python requirement but I am now updating to the later version of Python (previously on 3.7.1) - and will install latest version of galpy! On this note: Does the unpickling of snapshot potentials in the latest version depend on pynbody? The reason I ask is, pynbody can't be installed on Windows (I had an issue with this at least - it was due to a missing library that's specific to Unix). Is the old method of redefining the original potential using the pickled dictionary  (referencing https://github.com/jobovy/galpy/discussions/507) still valid for these cases?

**Further context**

- An example workflow of what I'm currently doing: I load in the public snapshot Auriga snapshot (from: https://wwwmpa.mpa-garching.mpg.de/auriga/gaiamock.html -  "Raw simulation snapshot data" -> link in the first paragraph) into pynbody
- I then separate its components into the stellar, gas, and DM components (done with the snapshot.s, snapshot.g, and snapshot.d conventions, where snapshot is the variable name allocated to the loaded simulation file).
- I used the spi snippet above for each component, created the potentials, and then pickled them using the method from #507. 
- Loaded in the pickled potentials into another script, unpickled them and reassigned the _origPot and _s variables. 
- Created a series of random vxvvs with an integration time of hubble time
- Ran the orbital instances in parallel by parsing the vxvvs for each binary within a single stacked array
- Orbits run fine with the included galpy potentials but run slowly with the interpolated potentials.

**Reply by jobovy** [reply]
Okay, just a quick question and comment to try to make progress quickly: Are you sure that the code is using the C extensions to do orbit integration? Are you getting any warnings saying that C integration can't be used? The only reason I can think of for the issue that you are having is that you are integrating in Python and some orbits going beyond the grid edge. In Python, the code then falls back onto the original potential (`SnapshotRZPotential` in this case) and evaluating that would be very slow, resulting in a very slow orbit integration. While if this is happening, you should expand the grid, really the thing you want to fix is the code not using C. If you are working on Windows, you need to make sure the GSL is installed and can be accessed by the code, because the galpy wheel does not include it (it does on Linux / Mac): https://docs.galpy.org/en/latest/installation.html.

On 3., looking at the code, it seems like you need `interpPot=True` to get the rotation curve, but you shouldn't need `interpepifreq` or `interpverticalfreq`.

Looks like pynbody doesn't support Windows. Yes, the old method of redefining the old potential should still work fine in your case.

**Reply by smandhai** [reply]
Thank you for the prompt response!

No C errors at the moment - it seems to be running with C from what I've seen! I'm also currently using Ubuntu (locally) but switch to Windows every so often depending on what I'm working on at the time. The point about the grid being too small might be worth exploring as some of the orbits I'm working with can very well exceed the grid's boundaries! I'll take a deeper look into this! Does this also effect objects passing through the centre of the grid (i.e. below the minimum radius defined for the grid).

GSL is installed on both Ubuntu/Windows installations that I work with - have tested galpy previously on these and they've worked without returning any warnings. 

I'm going to play with the orbit parameters a bit more and try to get a system that can do a few orbits within the grid - will see how this goes.

**Reply by jobovy** [reply]
Okay, if C is working, then the only plausible issue is something weird about the orbit integration which causes it to take a long time. One thing you can try to diagnose this is to run the integration with a fixed time step, because then your integration should take as long as the `interpRZPotential` version of `MWPotential2014` (or 3 times as long because you have three interpolated potentials). You can do this with, e.g.,
```
orb.integrate(...,dt=(ts[1]-ts[0])/100,method='symplec4_c`)
```
where `ts` is your time array and I've just chosen a fixed time step that is 1/100 of the time resolution of your output. You can't use `dop853_c` here, because it has an adaptive time step.

Some more comments:
* Looking at your grid specification above, it seems somewhat implausible that an orbit would go off the grid: `rgrid=(np.log10(1e-4),np.log10(1e3),101),logR=True,zgrid=(-60.,60.,101)`. But the grid spacing seems possibly way too large here. Note that these quantities are in internal units, so in units of ro=8 kpc if you are using the standard units. In R, you are going from 1e-4 x ro to 1e3 x ro in only 100 log increments. That's 7 orders of magnitude, so only 10 points per order of magnitude. If you are using ro=8 kpc, you are also going from 8 mpc out to 800 kpc. I doubt you need to resolve 7 orders of magnitude? If you do, you should make the grid spacing smaller. Similarly in z, you go from -60 x ro to 60 x ro in 101 steps, so in increments of ~1.2 x ro. Again, assuming that you use ro=8 kpc, that means that you don't resolve the disk. So either reduce the minimum/maximum z or increase the number of z grid points.
* Are you using the same grids for all three components? If yes, then there is no advantage to doing them separately, only a speed disadvantage.

**Reply by smandhai** [reply]
Ah thank you so much for both the possible tests and clarifications. I'll do these tests within the next few days and report back. As for the grid size, some of the systems that I'm working with can travel beyond 100 kpc and in some cases (<1%) can reach 1 Mpc - so ideally I'd need some way to track these. Although, if I recall, by defining "_origPot" and "_s" I can go beyond the grid, so that may not be so much of an issue (it doesn't need to be super well resolved at this distance).

**Reply by jobovy** [reply]
In C, the original potential isn't evaluated, so in C the force is frozen to be that at the boundary of the grid. If these are rare cases, probably the most efficient thing to do is to setup a grid that goes at less far and then test whether R > Rmax or |z| > zmax anywhere along an orbit. The few orbits that will do this could then be integrated in Python, which would switch to `origPot` outside the grid. 

Looking through the code, I also noticed that in C, the potential is always assumed to be symmetric around z=0. Maybe it shouldn't assume this, but it means that the negative part of your zgrid isn't doing anything. Do you think the z-asymmetry is important for your systems?

**Reply by smandhai** [reply]
z-symmetry might be important for I need to do, as ideally, I'm trying to capture features above and below the galactic plane that may affect the orbit of the stellar systems that I'll be injecting. As for the integration beyond the boundary, I'll keep a note of this! I'll likely try to do this for cases where the stellar systems become unbound from the galaxy.

**Reply by smandhai** [reply]
Was reproducing the potentials and came across the issue "Weird about the pickling not working in 1.8.1, but if it works again in 1.8.2 then I guess there is nothing to fix! Maybe the reason you can't upgrade is that your Python version is < 3.8? galpy 1.8.2 requires Python >= 3.8." on my laptop again with GalPy 1.8.2. 



I get the following traceback:

```
    dm_pot = pickle.load(f) <--- unpickling a pickled **pot.InterpSnapshotRZPotential** instance

  File "/home/*user directory*/anaconda3/lib/python3.8/site-packages/galpy/potential/SnapshotRZPotential.py", line 614, in __setstate__
    pdict['_s']= pynbody.new(star=len(pdict['_mass']))

KeyError: '_mass'
```

This is especially weird as it works fine on my other galpy 1.8.2 installation on another machine. I'm tempted to check the installation order of pynbody/galpy to see if that's affecting it. Might need to reinstall both with galpy first. (Should I submit this as an "issue"?)

Back to the speed update:
Using fixed time steps appears to be working - not sure about the accuracy just yet. I've created a combined interpolated potential and am in the process of checking the speed. I've reduced the grid size and increased the resolution. I'll report back once this has been checked.

**Reply by smandhai** [reply]
The combined interpolated potential works faster - it takes 10 orbits just under two minutes to finish. Using timesteps with a step size of 100, it takes around 2 seconds. I'll continue to play around but this is somewhat workable for now, as I won't be orbiting a large number of systems just yet. 

As I've reduced the gridsize, I'll work in the partial python orbit integration that you suggested! I'll need to figure out the most optimal way to do this, especially if the systems I'm working with leave and re-enter the grid several times.

**Reply by jobovy** [reply]
That pickling error is very strange, according to the code I don't see how `_mass` would not be available (https://github.com/jobovy/galpy/commit/6d866afbcba316f517657b916e45b223d55e9238). If you can create a simple reproducible example of the bug, feel free to open an issue. Of course, if you can't reliably reproduce it yourself, then it might be an issue with how you are using the code (perhaps you are using a pickle created after removing the `_s`?). 

If z-symmetry is important to you, you might want to open an issue about that. Should be an easy enough fix, although we'll need to come up with a good test (could add a small z offset bulge so there's a small z-asymmetry that would then be ignored by the C code).

In terms of leaving/re-entering the grid multiple times: I would just integrate any orbit that ever leaves the grid fully in Python, so not worry about resolving the multiple re-enters. Or is that too slow?

**Reply by smandhai** [reply]
Thank you for the response, Jo! For the pickling issue, I think it's a local issue on my laptop as the exact same script with the exact same pickled objects run fine on the work machines I'm using. I'll tinker with this and find a solution, I'll let you know if there's a fix in case someone else stumbles into the same issue.

I haven't tested the Python integration but I suspect that might be fine to be honest, especially as I assume most objects will take a while to re-enter the grid once they exit. 

I'll revisit the z-symmetry issue at a later point as I'm not at a point where I can determine if its impact is huge. I'll think of a test nonetheless and compare the offsets to the results I get with it as it currently is. 

For now, I think the adaptive timesteps are working well and the combined potentials (so I think as far as the optimisation in the original question go, this has been resolved!). 

I'll continue working on the different parts and add to this thread when I've tested the ideas.

**Reply by smandhai** [reply]
Related to the  `_mass` error, some additional notes:

Clean anaconda reinstall didn't solve the issue. I realised that the pickled potential was created in 1.8.0, and therefore needed to have `_s` removed before pickling to get around the Threadlock issue in #509. I tried to create the pickles for the potentials in 1.8.2 but this returns a `AttributeError: 'NoneType' object has no attribute 'samples_to_coefficients'` (probably owing to the fact that `_s` and `_origPot` are removed. I'll try adding these back in and see what happens). 

It's still baffling why the scripts and pickles run perfectly fine on a venv environment with galpy 1.8.2 installed on the work machine. The only other explanation I can think of is that the work machine is defaulting to galpy 1.8.0 and ignoring the venv installation (I've checked both via pip to confirm their versions). 

**Edit**: Suspicions confirmed - I loaded ipython within the virtual environment and the galpy version reported (`galpy.__version__`) is 1.7.2 (version on the base installation of python on my work machine) despite pip reporting 1.8.2. Not sure why this is but it would explain why there's a discontinuity between different machines. I suspect my laptop had a true 1.8.2 version! So I suspect the issue in the end was an incompatibility with pickling pre-1.8.2 snapshots and loading them in via 1.8.2. I'll do the final test where I add `_s `and `_origPot` back in and see if I can get everything working on 1.8.2.

**Edit 2**: Indeed, the `_mass` error does occur on the second machine when I force the usage of the usage of version 1.8.2. So that reaffirms the incompatibility issue!

**Reply by jobovy** [reply]
Okay, thanks for checking! So it seems like there is no intrinsic issue with the pickling then 👍
